DEVELOPING.md

```diff
@@ -1,9 +1,138 @@
-# POM Manipulation Extension (PME) Release Process
 
-This document describes how to setup and perform a new POM Manipulation
-Extension (PME) release.
+# Table of Contents
 
-## Conventions
+<!-- TocDown Begin -->
+  * [Prerequisites](#prerequisites)
+  * [Coding](#coding)
+    * [Exception Handling](#exception-handling)
+    * [Configuration](#configuration)
+  * [Style & Copyright](#style--copyright)
+  * [IntelliJ](#intellij)
+  * [Documentation](#documentation)
+  * [Compiling](#compiling)
+  * [Testing](#testing)
+    * [Code Coverage](#code-coverage)
+  * [Release Process](#release-process)
+    * [Conventions](#conventions)
+    * [Release Setup](#release-setup)
+    * [Pre-release Checks](#pre-release-checks)
+    * [Run the Release](#run-the-release)
+    * [Roll Back a Release](#roll-back-a-release)
+<!-- TocDown End -->
+
+
+## Prerequisites
+
+* Java 1.8 or later
+* Maven 3.3.9 or later (to build, it requires 3.1 or later to run)
+
+## Coding
+
+### Exception Handling
+
+The three built in Exceptions all support SLF4J style `{}` substitution parameters
+
+    ManipulationException
+    RestException
+    ManipulationUncheckedException
+
+For example
+
+    throw new ManipulationException( "Unable to detect charset for file {}", jsonFile, exceptionObject );
+    throw new ManipulationException( "Internal project dependency resolution failure ; replaced {} by {}", old, d );
+    throw new ManipulationException( "Unable to parse groovyScripts", exceptionObject );
+
+### Configuration
+
+Any configuration property value **must** have a static String together with the internal configuration annotation e.g.
+```
+    @ConfigValue( docIndex = "project-version-manip.html#manual-version-suffix" )
+    public static final String VERSION_SUFFIX_SYSPROP= "versionSuffix";
+```
+The `docIndex` value represents the property index for the gh-pages. This is automatically generated upon build inside
+`root/target/property-index-subset.md`. A new class `org.commonjava.maven.ext.core.ConfigList` is also generated that contains
+a HashMap of all properties ; this is used for validation input checking within the CLI /  EXT.
+
+
+## Style & Copyright
+
+Eclipse compatible `codestyle.xml` and `eclipse.importorder` files are supplied inside the `ide-config` directory which
+may also be imported into IntelliJ via the EclipseCodeFormatter. There is also an IntelliJ compatible copyright template
+suitable for use when the project is imported into IntelliJ.
+
+The `.idea` folder contains a copyright template suitable for use when the project is imported into IntelliJ.
+
+## IntelliJ
+
+The following plugin is required:
+
+ * https://plugins.jetbrains.com/plugin/6317-lombok/
+
+It is also recommended to install:
+
+ * https://plugins.jetbrains.com/plugin/6546-eclipse-code-formatter
+    * (which allows importing of the code style)
```

Comment:
Should we mention or link to [Eclipse Code Formatter](https://github.com/krasa/EclipseCodeFormatter#instructions) on how to set up the code formatter? Unfortunately, I seem to get different results between that and [formatter-maven-plugin](https://github.com/revelc/formatter-maven-plugin).

Reply:
I will add a link. Probably investigating `.editorconfig` files might be a good way forward but not relevant I think for this ticket.